pkg/handler/collector/oci/oci.go

```diff
@@ -209,63 +208,87 @@ func (o *ociCollector) fetchOCIArtifacts(ctx context.Context, repo string, rc *r
 	}
 
 	digest := manifest.GetDigest(m)
+
 	digestFormatted := fmt.Sprintf("%v-%v", digest.Algorithm(), digest.Encoded())
 	suffixList := []string{"att", "sbom"}
 	for _, suffix := range suffixList {
 		digestTag := fmt.Sprintf("%v.%v", digestFormatted, suffix)
 		// check to see if the digest + suffix has already been collected
 		if !contains(o.checkedDigest[repo], digestTag) {
 			imageTag := fmt.Sprintf("%v:%v", repo, digestTag)
-			r, err := ref.New(imageTag)
+			err = fetchOCIArtifactBlobs(ctx, rc, imageTag, docChannel)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed retrieving artifact blobs from registry fallback artifacts: %w", err)
 			}
+			o.checkedDigest[repo] = append(o.checkedDigest[repo], digestTag)
+		}
+	}
 
-			// if `.att` or `.sbom`` do not exist for specified digest
-			// log error and continue
-			m, err = rc.ManifestGet(ctx, r)
-			if err != nil {
-				// this is a normal behavior, not an error when the digest does not have an attestation
-				// explicitly logging it as info to avoid call-stack when logging
-				logger.Infof("unable to get manifest for %v: %v", imageTag, err)
-				continue
-			}
+	referrerOpts := []scheme.ReferrerOpts{}
+	referrerOpts = append(referrerOpts, scheme.WithReferrerAT("application/spdx+json"))
 
-			// go through layers in reverse
-			mi, ok := m.(manifest.Imager)
-			if !ok {
-				return fmt.Errorf("reference is not a known image media type")
-			}
-			layers, err := mi.GetLayers()
-			if err != nil {
-				return err
-			}
-			for i := len(layers) - 1; i >= 0; i-- {
-				if ctx.Err() != nil {
-					return ctx.Err()
-				}
-				blob, err := rc.BlobGet(ctx, r, layers[i])
-				if err != nil {
-					return fmt.Errorf("failed pulling layer %d: %w", i, err)
-				}
-				btr1, err := blob.RawBody()
-				if err != nil {
-					return err
-				}
+	referrerList, err := rc.ReferrerList(ctx, image, referrerOpts...)
+	if err != nil {
+		return err
+	}
+	for _, referrerDesc := range referrerList.Descriptors {
+		imageTag := fmt.Sprintf("%v@%v", repo, referrerDesc.Digest.String())
+		err = fetchOCIArtifactBlobs(ctx, rc, imageTag, docChannel)
+		if err != nil {
+			return err
+		}
+	}
 
-				doc := &processor.Document{
-					Blob:   btr1,
-					Type:   processor.DocumentUnknown,
-					Format: processor.FormatUnknown,
-					SourceInformation: processor.SourceInformation{
-						Collector: string(OCICollector),
-						Source:    imageTag,
-					},
-				}
-				docChannel <- doc
-			}
-			o.checkedDigest[repo] = append(o.checkedDigest[repo], digestTag)
+	return nil
+}
+
+func fetchOCIArtifactBlobs(ctx context.Context, rc *regclient.RegClient, artifact string, docChannel chan<- *processor.Document) error {
```

Comment:
A suggestion is to use non-blocking calls to get these data in parallel instead of sequentially.

The BlogGet is an HTTP method that can utilize go routines to have non-blocking calls.

Here is an example of how it can be done. 

``` golang
func fetchOCIArtifactBlobs(ctx context.Context, rc *regclient.RegClient, artifact string, docChannel chan<- *processor.Document) error {
    logger := logging.FromContext(ctx)
    r, err := ref.New(artifact)
    if err != nil {
        return err
    }

    m, err := rc.ManifestGet(ctx, r)
    if err != nil {
        // this is a normal behavior, not an error when the digest does not have an attestation
        // explicitly logging it as info to avoid call-stack when logging
        logger.Infof("unable to get manifest for %v: %v", artifact, err)
        return nil
    }

    // go through layers in reverse
    mi, ok := m.(manifest.Imager)
    if !ok {
        return fmt.Errorf("reference is not a known image media type")
    }
    layers, err := mi.GetLayers()
    if err != nil {
        return err
    }

    // Create a channel to collect errors from goroutines
    errorChannel := make(chan error, len(layers))

    // Use a WaitGroup to wait for all goroutines to finish
    var wg sync.WaitGroup

    for i := len(layers) - 1; i >= 0; i-- {
        if ctx.Err() != nil {
            return ctx.Err()
        }

        // Increment the WaitGroup counter
        wg.Add(1)

        // Start a goroutine to fetch the data and handle errors
        go func(layerIndex int) {
            defer wg.Done() // Decrement the WaitGroup counter when done

            blob, err := rc.BlobGet(ctx, r, layers[layerIndex])
            if err != nil {
                // Send the error to the errorChannel
                errorChannel <- fmt.Errorf("failed pulling layer %d: %w", layerIndex, err)
                return
            }

            btr1, err := blob.RawBody()
            if err != nil {
                // Send the error to the errorChannel
                errorChannel <- err
                return
            }

            doc := &processor.Document{
                Blob:   btr1,
                Type:   processor.DocumentUnknown,
                Format: processor.FormatUnknown,
                SourceInformation: processor.SourceInformation{
                    Collector: string(OCICollector),
                    Source:    artifact,
                },
            }

            // Send the document to the docChannel
            docChannel <- doc
        }(i)
    }

    // Wait for all goroutines to finish
    wg.Wait()

    // Close the errorChannel to signal that all errors have been collected
    close(errorChannel)

    // Check if any errors occurred during processing
    for err := range errorChannel {
        if err != nil {
            return err // Return the first error encountered
        }
    }

    return nil
}
```

Reply:
Agreed, this is a good improvement to the layer downloads. I will add this.

Reply:
Thanks, @ridhoq!

Reply:
I went ahead and added this suggestion to a higher level instead of the layer pulls as in most cases we will only have one layer. We are now concurrently grabbing across manifest lists and referrers.